### PR TITLE
Remove chpl_task_yield from meteor-parallel-alt's atomic lock/unlock

### PR DESCRIPTION
**Pull request by ronawho**
This is basically a duplicate of #4226. #4226 had to be backed out because of a
deadlock under our memory tracking testing. This skips meteor-parallel-alt for
memory tracking and adds a futre to catpure the problem. The basic issue is
that the atomic lock/unlock requires that tasks aren't yielding. Memory tracking
currently uses a sync var as a lock, and sync vars can occasionally yield
(depending on the tasking layer), which causes deadlock.

### CODE REVIEW

**Comment by ronawho** [comment]
@gbtitus @ben-albrecht @bradcray look ok?

Note that I used `{ }` instead of `do ;` for atomic lock because I thought it looked a little cleaner, but I go can back to the other style if you prefer.

**Comment by gbtitus** [comment]
This makes me _really_ nervous.  We've gone to some trouble in the past to avoid deadlocks and livelocks caused by spin-waiting without yielding.  (The chpl_task_yield() in Atomics.chpl:waitFor() is a good example.)  memTracking is far from the only place we could end up encountering a sync lock that broke this code or anything like it.  In fact, any user could easily write something just like what we're dealing with here, with the sync hidden in their own code or some library they're not familiar with, instead of buried in memTracking.  The comment in memTrackDeadlock.chpl helps a little.  At a minimum we should add a corresponding comment to meteor-parallel-alt.chpl saying that normal coding practice requires the yield to avoid the potential for livelock/deadlock, that it was removed for performance, and that if this code or anything based on it hangs, the first step should be to add the yield back in and see if it works right.

File: test/memory/elliot/memTrack/memTrackDeadlock.future [hunk 1]
```diff
@@ -0,0 +1,9 @@
+bug: memTracking deadlocks for tests that can't task yield
+
+memory tracking currently uses a sync var as a lock. sync vars can yield the
+current ask and for programs that are relying on their tasks not yielding this
```
Comment (ben-albrecht) [inline test/memory/elliot/memTrack/memTrackDeadlock.future]: 'current ask' -> 'current task'

File: test/memory/elliot/memTrack/memTrackDeadlock.future [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+can lead to deadlock. mem tracking really shouldn't be using sync vars then,
+and perhaps should using a pthread_mutex intead (although then we have to be
+careful that we lock/unlock with the same pthread.)
+
+When this future is retired, please remove the skipif
```
Comment (ben-albrecht) [inline test/memory/elliot/memTrack/memTrackDeadlock.future]: Should we reference meteor benchmark in this future, or is meteor's reference to this future enough? i.e. when this future gets retired, will the retirer be able to figure out they can remove meteor's `skipif` as well?

**Comment by ben-albrecht** [comment]
Looks good. Thanks for taking care of this.

> Note that I used { } instead of do ; for atomic lock because I thought it looked a little cleaner, but I go can back to the other style if you prefer.

I agree it looks better. Admittedly, I code golfed a little with this submission, since it was our fastest version and far from anything I'd call 'elegant'. However, this change seems worth the delta gz.

**Comment by ronawho** [comment]
@gbtitus yeah, you have some good points. While I don't really want to add comments to meteor (since we'll be submitting it for the world to see) I also don't want to put that code out there as a "recommended" way to code.

What do we think of a compromise. Instead of the current yield every contested lock, or this never yield, what about if we yield every so often. It's a little more code, but it's not too terrible: https://github.com/chapel-lang/chapel/compare/master...ronawho:alternative-meteor-yield

Let me know what you think. That branch gets us the same performance as this PR and the number of iterations before yielding was basically the smallest value that avoids any yields on the shootout box.

**Comment by gbtitus** [comment]
Ah, that I am much happier with.  In fact, I think that can actually hold it's own as a "recommended" way to code, although one further step might be to encapsulate it into a standard "locks" module or some such.  I would recommend 2 changes, though: augment the comment to mention that it is coded that way because spinning without yielding is not Chapelriffic (sp?), and name the lock bool something like "lock" instead of "a".  :-)

**Comment by ronawho** [comment]
Unfortunately the number of lockAttempts to avoid yielding on the shootout machine is about `2**16`, which ends up making `--memTrack` take 300 seconds (compared to .2 seconds with yielding every time.)

We really need some way to avoid yielding at all for performance runs, but yield frequently when we need to for things like memTrack.

**Comment by gbtitus** [comment]
Check memTracking.memTrack and behave differently when it's on or off? That's admittedly ugly, but when we say that Chapel's multiresolution design allows programming at higher levels for abstraction or lower levels for performance, it's implicit that coding at low levels may require one to do ugly things.  (The high-level way to code this part of meteor-parallel-alt is to use a sync var for the lock, of course.)

**Comment by ronawho** [comment]
Checking memTracking.memTrack seems like a bad idea to me. Like you said, there could be future reasons for something introducing a yield beyond memtracking, so checking memLeaks would only be a bandaid 

I opted to go for a backoff spin lock. It basically yield infrequently at first, and then yields more and more frequenly until it yields every iteration. I think something like https://github.com/chapel-lang/chapel/compare/master...ronawho:backoff-spin-lock-meteor will work. It still gets us the best time on the shootout box, and it finishes memLeaks in .2 seconds

**Comment by gbtitus** [comment]
Ah, I misunderstood the problem a little bit in my previous response.  This looks much better than what I proposed.  Though the question "Why divide the maxLockAttempts by 4 each time we back off?" occurred to me right away.  Does that behave significantly better than dividing it by 2, for example?  (I realize you probably didn't do a full search of the parameter space, but the >>=2 does look like a specifically chosen value.)

**Comment by ronawho** [comment]
`>>=2` wasn't specifically chosen, it was mostly a random choice and it happened to work well. `>>=1` performs as well so I'll switch to that.

**Comment by ronawho** [comment]
Replaced by https://github.com/chapel-lang/chapel/pull/4300

**Pull request closed without merging**